packages/react-admin/src/AdminRoutes.js

```diff
@@ -1,28 +1,39 @@
-import React, { Component } from 'react';
+import React, { Children, Component } from 'react';
 import PropTypes from 'prop-types';
-import { Redirect, Route, Switch } from 'react-router-dom';
+import { Redirect, Route, Switch, withRouter } from 'react-router-dom';
 import { connect } from 'react-redux';
 import compose from 'recompose/compose';
 import getContext from 'recompose/getContext';
 
 import CrudRoute from './CrudRoute';
 import NotFound from './mui/layout/NotFound';
 import Restricted from './auth/Restricted';
-import { AUTH_GET_PERMISSIONS } from './auth';
-import { declareResources as declareResourcesAction } from './actions';
+import {
+    declareResources as declareResourcesAction,
+    userCheck as userCheckAction,
+} from './actions';
+import { getPermissions } from './reducer/admin/auth';
 
 export class AdminRoutes extends Component {
     componentDidMount() {
-        return this.initializeResources(this.props.children);
+        this.getPermissions();
     }
 
-    async initializeResources(children) {
-        if (typeof children === 'function') {
-            let permissions;
-            if (this.props.authClient) {
-                permissions = await this.props.authClient(AUTH_GET_PERMISSIONS);
-            }
+    async componentWillReceiveProps(nextProps) {
+        if (nextProps.permissions !== this.props.permissions) {
+            await this.initializeResources(nextProps);
+        }
+    }
 
+    getPermissions = () => {
+        const { userCheck, location, match } = this.props;
+        userCheck({}, location && location.pathname, match && match.params);
```

Comment:
Am I understanding correctly that each visit to a page triggers an AUTH_CHECK (the same check that at login) instead of AUTH_GET_PERMISSIONS? If so, I think it’s not a good idea. We should allow a separate implementation of authentication and authorization in the auth client

Reply:
Yes, each location change will trigger an AUTH_CHECK. I'm not sure about the benefits of such separation. Can you explain ?